tests/constrainedExample.h

```diff
@@ -0,0 +1,90 @@
+/* ----------------------------------------------------------------------------
+ * GTDynamics Copyright 2020, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * See LICENSE for the license information
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file  constrainedExample.h
+ * @brief Examples for constrained optimization. From
+ * http://www.seas.ucla.edu/~vandenbe/133B/lectures/nllseq.pdf
+ * @author Yetong Zhang
+ */
+
+#pragma once
+
+#include <gtsam/base/Matrix.h>
+#include <gtsam/base/Vector.h>
+#include <gtsam/inference/Symbol.h>
+#include <gtsam/nonlinear/ExpressionFactor.h>
+#include <gtsam/nonlinear/NonlinearFactor.h>
+#include <gtsam/nonlinear/expressions.h>
+
+namespace gtdynamics {
+namespace constrained_example {
+
+using gtsam::Double_;
+using gtsam::Vector2;
+using gtsam::Vector2_;
+using gtsam::Key;
+using gtsam::Symbol;
+using gtsam::Expression;
+using gtsam::Vector1;
+
+/** First cost function. */
+double cost1(const double &x1, const double &x2,
+             gtsam::OptionalJacobian<1, 1> H_x1 = boost::none,
+             gtsam::OptionalJacobian<1, 1> H_x2 = boost::none) {
+  double result = x1 + exp(-x2);
+  if (H_x1) H_x1->setConstant(1.0);
+  if (H_x2) H_x2->setConstant(-exp(-x2));
+  return result;
+}
+
+/** Second cost function. */
+double cost2(const double &x1, const double &x2,
+             gtsam::OptionalJacobian<1, 1> H_x1 = boost::none,
+             gtsam::OptionalJacobian<1, 1> H_x2 = boost::none) {
+  double result = x1 * x1 + 2 * x2 + 1;
+  if (H_x1) H_x1->setConstant(2 * x1);
+  if (H_x2) H_x2->setConstant(2.0);
+  return result;
+}
+
+/** Constraint function g(x). */
+double constraint1(const double &x1, const double &x2,
```

Comment:
This one can be done entirely in expressions with the overloaded operators, I think